api/utils/sshutils/ssh.go

```diff
@@ -264,7 +264,7 @@ func hostKeyFallbackFunc(knownHosts []ssh.PublicKey) func(hostname string, remot
 
 // KeysEqual is constant time compare of the keys to avoid timing attacks
 func KeysEqual(ak, bk ssh.PublicKey) bool {
-	a := ssh.Marshal(ak)
-	b := ssh.Marshal(bk)
+	a := ssh.MarshalAuthorizedKey(ak)
+	b := ssh.MarshalAuthorizedKey(bk)
 	return (len(a) == len(b) && subtle.ConstantTimeCompare(a, b) == 1)
```

Comment:
I realize this has always been this way, but the `len` check is actually not necessary here, `ConstantTimeCompare` will take care of it.

Reply:
Or to avoid base64 encoding timing differences:

```go

// KeysEqual is constant time compare of the keys to avoid timing attacks
func KeysEqual(ak, bk ssh.PublicKey) bool {
	if ak.Type() != bk.Type() {
		return false
	}

	a := ak.Marshal()
	b := bk.Marshal()
	return subtle.ConstantTimeCompare(a, b) == 1
}
```

Reply:
Ah, TIL - the key string re-encodes the type. It's redundant. This is fine.